tests/yast2_gui/yast2_storage_ng.pm

```diff
@@ -0,0 +1,218 @@
+    sleep 1;
+    # custom size
+    send_key "alt-u";
+    sleep 1;
+    # select entry and type partition size
+    send_key "alt-s";
+    wait_screen_change { type_string "170MiB" };
+    wait_screen_change { send_key(is_sle('<=12-sp4') ? "alt-o" : "alt-n") };
+
+    assert_screen "yast2_storage_ng-partition-resized";
+    wait_screen_change { send_key "alt-n" };
+    sleep 1;
+    wait_screen_change { send_key "alt-f" };
+
+    ### DELETE PARTITION ###
+    sleep 3;
+    start_y2sn $self;
+    select_vdb;
+    wait_screen_change { send_key "alt-l" };
+    wait_screen_change { send_key "alt-y" };
+    assert_screen "yast2_storage_ng-unpartitioned";
+    wait_screen_change { send_key "alt-n" };
+    sleep 1;
+    wait_screen_change { send_key "alt-f" };
+
+    ### CREATE VOLUME GROUP ###
+    sleep 3;
+    start_y2sn $self;
+    assert_and_click "yast2_storage_ng-select-vol-management";
+    wait_screen_change { send_key "alt-a" };
+    # alt-v doesn't work reliably, so we have to use assert_and_click
+    assert_and_click "yast2_storage_ng-add-volume-group" if is_sle;
+    wait_screen_change { type_string "vgtest" };
+    assert_and_click "yast2_storage_ng-vg-select-device";
+    send_key "alt-a";
+    wait_screen_change { send_key(is_sle('<=12-sp4') ? "alt-f" : "alt-n") };
+
+    #  go to system view
+    send_key "alt-s";
+    assert_and_dclick "yast2_storage_ng-select-vgtest";
+
+    my $fs_page_shortcut = is_sle("<=12-sp4") ? "alt-f" : "alt-n";
+    wait_screen_change { send_key "alt-i" } if is_opensuse;
+
+    # XFS, non encrypted
+    add_logical_volume "lv1", "alt-d";
+    wait_screen_change { send_key $fs_page_shortcut };
+
+    # EXT4, encrypted
+    add_logical_volume "lv2", "alt-d";
+    assert_and_click "yast2_storage_ng-filesystem-dropdown";
+    send_key_until_needlematch("yast2_storage_ng-ext4", "up");
+    send_key "ret";
+    encrypt_partition;
+
+    # BtrFS, encrypted
+    add_logical_volume "lv3", "alt-d";
+    assert_and_click "yast2_storage_ng-filesystem-dropdown";
+    send_key_until_needlematch("yast2_storage_ng-btrfs", "up");
+    send_key "ret";
+    # BtrFS encryption fails on SLE 12.x, therefore we skip it
+    if (is_sle("<=12-sp4")) {
+        wait_screen_change { send_key $fs_page_shortcut };
+    } else {
+        encrypt_partition;
+    }
+
+    # Raw, non encrypted
+    my $raw_shortcut = is_sle("<=12-sp4") ? "alt-a" : "alt-r";
+    add_logical_volume "lv4", $raw_shortcut;
+    # on SLE 12.x the alt-a shortcut doesn't seem to work reliably, so here we 'force' the raw format
+    send_key "alt-d" if is_sle("<=12-sp4");
+    sleep 1;
+    wait_screen_change { send_key $fs_page_shortcut };
+
+    # summary and finish
+    wait_screen_change { send_key "alt-n" };
+    sleep 1;
+    wait_screen_change { send_key "alt-f" };
+
+    # check that all logical volumes have been created
+    sleep 5;
+    select_console "root-console";
+    assert_script_run 'for i in {1..4}; do lvdisplay "/dev/vgtest/lv${i}"; done';
```

Comment:
Just a general comment. I know that it's the thing we have in many other places, but for the new tests we need to improve the situation. As of now this module is mainly automating some UI interaction. I strongly suggest not only click menus, but validate that YaST module has done thing correctly. So what's about extending this part to check more things. Also, in case of cycle, I believe that exit code will be zero if first 3 iterations will fail and 4-th will succeed.
Try following:
1. Create directory `test1`
2. Run:
```bash
for i in {1..2}; do mkdir test${i}; done
```
3. Exit code is zero, because last command was successful, but first iteration fails as directory already exists.

Reply:
I'm actually checking whether things went right (partition created, deleted and resized) but I just noticed that I'm checking it before confirming the action, thus making the check useless. I need to change the order of the events!

Good catch about the exit code! Haven't thought about that, I'll fix it right now :+1:

Reply:
If you have any suggestion on what else I could check please let me know. There's always room for improvement :smile:

Reply:
My main recommendation would be not to use UI of the tool under test to validate results, as you've mentioned above. For example, in the test you resize partition and check if it got resized in the UI, which is still useful check, but we don't assure that partition will be resized in the SUT. There is still risk that fdisk has a bug and shows resized partition, but validating tool under test by itself is not the best choice in case there are alternatives.

So for the checks, I would simply add steps to apply changes and validate them similarly to what you've done with logical volumes.

Reply:
Right, that's really a good point! Ok, I'll implement this :+1:

Reply:
Awesome! As I've said, unfortunately we don't have good guidelines and most of the existing tests are not serving a good example =)